File: services/keyword-analysis/kw_rank/utils/error_handler.py
"""
Centralized error handling utilities for Python keyword analysis service.

This module provides standardized error handling, logging, and result creation
to ensure consistent error management across the Python codebase. Mirrors the
JavaScript ErrorHandler class for cross-language consistency.
"""

import sys
import traceback
from pathlib import Path
from typing import Any, Dict, List, Optional, Union
from .error_types import ErrorTypes, ErrorSeverity, ContextTypes, ERROR_CONFIG, LEGACY_ERROR_MAPPING


class ErrorHandler:
    """Centralized error handling and logging utilities."""
    
    @staticmethod
    def log_error(message: str, error: Optional[Exception] = None, 
                  context: Optional[Dict[str, Any]] = None, 
                  details: Optional[List[str]] = None) -> None:
        """
        Log error with structured formatting and context.
        
        Args:
            message: Primary error message
            error: Original exception object (optional)
            context: Additional context information (optional)
            details: List of additional details (optional)
        """
        if not ERROR_CONFIG['logging']['enabled']:
            return
        
        # Format primary error message with emoji
        emoji = ERROR_CONFIG['emojis']['error'] if ERROR_CONFIG['formatting']['use_emojis'] else ''
        print(f"{emoji} {message}")
        
        # Log original error message if provided
        if error and str(error):
            print(f"   Original error: {str(error)}")
        
        # Log context information if provided and enabled
        if context and ERROR_CONFIG['formatting']['include_context']:
            for key, value in context.items():
                if value is not None:
                    print(f"   {key}: {value}")
        
        # Log additional details if provided
        if details:
            max_lines = ERROR_CONFIG['logging']['max_context_lines']
            for i, detail in enumerate(details[:max_lines]):
                print(f"   • {detail}")
            
            if len(details) > max_lines:
                print(f"   ... and {len(details) - max_lines} more details")
        
        # Log stack trace if enabled and error provided
        if (error and ERROR_CONFIG['formatting']['include_stack_trace']):
            print("   Stack trace:")
            traceback.print_exc()
    
    @staticmethod
    def log_app_error(message: str, component: str, error: Optional[Exception] = None,
                      context: Optional[Dict[str, Any]] = None) -> None:
        """
        Log application-level error with component context.
        
        Args:
            message: Error message
            component: Component or module name where error occurred
            error: Original exception (optional)
            context: Additional context (optional)
        """
        app_context = {'component': component}
        if context:
            app_context.update(context)
        
        ErrorHandler.log_error(message, error, app_context)
    
    @staticmethod
    def log_service_error(message: str, service: str, operation: str,
                         error: Optional[Exception] = None,
                         context: Optional[Dict[str, Any]] = None) -> None:
        """
        Log service-level error with structured context.
        
        Args:
            message: Error message
            service: Service name
            operation: Operation being performed
            error: Original exception (optional)
            context: Additional context (optional)
        """
        service_context = {
            'service': service,
            'operation': operation
        }
        if context:
            service_context.update(context)
        
        ErrorHandler.log_error(message, error, service_context)
    
    @staticmethod
    def create_result(success: bool, data: Any = None, error_message: str = None,
                     error_type: str = ErrorTypes.UNKNOWN_ERROR, 
                     details: List[str] = None,
                     legacy_error_type: str = None) -> Dict[str, Any]:
        """
        Create standardized result object for consistent return values.
        
        Args:
            success: Whether the operation succeeded
            data: Result data for successful operations
            error_message: Error message for failed operations
            error_type: Standardized error type
            details: Additional error details
            legacy_error_type: Legacy error type for backward compatibility
        
        Returns:
            Standardized result dictionary
        """
        if success:
            return {
                'is_valid': True,
                'success': True,
                'data': data
            }
        else:
            result = {
                'is_valid': False,
                'success': False,
                'error': error_message,
                'error_type': error_type,
                'details': details or []
            }
            
            # Add legacy error type for backward compatibility if provided
            if legacy_error_type:
                result['legacy_error_type'] = legacy_error_type
            
            return result
    
    @staticmethod
    def validate_input(validator_func, input_data: Any, 
                      error_message: str = "Input validation failed",
                      expected_format: str = None) -> Dict[str, Any]:
        """
        Validate input using provided validator function with structured error reporting.
        
        Args:
            validator_func: Function to validate input (should return bool)
            input_data: Data to validate
            error_message: Custom error message
            expected_format: Description of expected format
        
        Returns:
            Standardized result object
        """
        try:
            is_valid = validator_func(input_data)
            
            if is_valid:
                return ErrorHandler.create_result(True, input_data)
            else:
                details = []
                if expected_format:
                    details.append(f"Expected format: {expected_format}")
                
                return ErrorHandler.create_result(
                    False, None, error_message, 
                    ErrorTypes.VALIDATION_ERROR, details
                )
        
        except Exception as e:
            ErrorHandler.log_error(f"Validator function failed: {str(e)}", e)
            return ErrorHandler.create_result(
                False, None, f"Validation error: {str(e)}", 
                ErrorTypes.VALIDATION_ERROR
            )
    
    @staticmethod
    def assert_required(value: Any, field_name: str) -> None:
        """
        Assert that a required value is present and not empty.
        
        Args:
            value: Value to check
            field_name: Name of the field for error reporting
        
        Raises:
            ValueError: If value is None, empty string, or empty collection
        """
        if value is None:
            raise ValueError(f"Required field '{field_name}' is None")
        
        if isinstance(value, str) and not value.strip():
            raise ValueError(f"Required field '{field_name}' is empty")
        
        if hasattr(value, '__len__') and len(value) == 0:
            raise ValueError(f"Required field '{field_name}' is empty")
    
    @staticmethod
    def build_file_context(file_path: str, additional_info: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Build file context for error reporting.
        
        Args:
            file_path: Path to the file
            additional_info: Additional context information
        
        Returns:
            File context dictionary
        """
        context = {
            'context_type': ContextTypes.FILE,
            'file_path': str(file_path),
            'file_exists': Path(file_path).exists() if file_path else False
        }
        
        if file_path and Path(file_path).exists():
            path_obj = Path(file_path)
            context.update({
                'file_size': path_obj.stat().st_size,
                'file_extension': path_obj.suffix
            })
        
        if additional_info:
            context.update(additional_info)
        
        return context
    
    @staticmethod
    def build_service_context(service_name: str, operation: str,
                             additional_info: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Build service context for error reporting.
        
        Args:
            service_name: Name of the service
            operation: Operation being performed
            additional_info: Additional context information
        
        Returns:
            Service context dictionary
        """
        context = {
            'context_type': ContextTypes.SERVICE,
            'service': service_name,
            'operation': operation
        }
        
        if additional_info:
            context.update(additional_info)
        
        return context
    
    @staticmethod
    def build_validation_context(field_name: str, expected_format: str = None,
                                actual_value: Any = None,
                                additional_info: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Build validation context for error reporting.
        
        Args:
            field_name: Name of the field being validated
            expected_format: Expected format description
            actual_value: Actual value that failed validation
            additional_info: Additional context information
        
        Returns:
            Validation context dictionary
        """
        context = {
            'context_type': ContextTypes.VALIDATION,
            'field_name': field_name
        }
        
        if expected_format:
            context['expected_format'] = expected_format
        
        if actual_value is not None:
            context['actual_value'] = str(actual_value)[:100]  # Truncate long values
        
        if additional_info:
            context.update(additional_info)
        
        return context
    
    @staticmethod
    def get_legacy_error_type(error_type: str, context: str = None) -> str:
        """
        Get legacy error type for backward compatibility.
        
        Args:
            error_type: Modern error type
            context: Context to help determine specific legacy type
        
        Returns:
            Legacy error type string
        """
        if error_type in LEGACY_ERROR_MAPPING:
            legacy_mapping = LEGACY_ERROR_MAPPING[error_type]
            if context and context in legacy_mapping:
                return context
            # Return first available legacy type as fallback
            return list(legacy_mapping.keys())[0] if legacy_mapping else error_type
        
        return error_type
    
    @staticmethod
    def exit_with_error(message: str, error_type: str = ErrorTypes.UNKNOWN_ERROR,
                       exit_code: int = 1) -> None:
        """
        Log error and exit the application.
        
        Args:
            message: Error message to log
            error_type: Type of error
            exit_code: Exit code (default: 1)
        """
        ErrorHandler.log_error(message)
        sys.exit(exit_code)

File: utils/error-handler.js
/**
 * Centralized Error Handler
 * Provides consistent error logging, formatting, and result creation across the application
 */

import { ERROR_TYPES, ERROR_SEVERITY, CONTEXT_TYPES } from './error-types.js';
import theme from '../theme.js';

/**
 * Configuration for error handling behavior
 */
const ERROR_CONFIG = {
  formatting: {
    useEmojis: true,
    includeContext: true,
    maxContextLines: 5,
    includeStackTrace: process.env.NODE_ENV === 'development'
  },
  logging: {
    logToConsole: true,
    logLevel: 'error'
  }
};

/**
 * Centralized Error Handler Class
 */
class ErrorHandler {
  /**
   * Core error logging with structured output
   * @param {Object} config - Error configuration object
   * @param {string} config.message - Primary error message
   * @param {Error} [config.error] - Original error object
   * @param {string} [config.type] - Error type from ERROR_TYPES
   * @param {string} [config.severity] - Error severity from ERROR_SEVERITY
   * @param {Array} [config.details] - Additional error details
   * @param {Object} [config.context] - Error context information
   */
  static logError({
    message,
    error = null,
    type = ERROR_TYPES.UNKNOWN_ERROR,
    severity = ERROR_SEVERITY.MEDIUM,
    details = [],
    context = {}
  }) {
    if (!ERROR_CONFIG.logging.logToConsole) return;

    const emoji = ERROR_CONFIG.formatting.useEmojis ? theme.messages.emojis.error : '';
    console.error(`${emoji} ${message}`);

    // Log original error message if provided
    if (error && error.message) {
      console.error(`   ${error.message}`);
    }

    // Log context information
    if (ERROR_CONFIG.formatting.includeContext && Object.keys(context).length > 0) {
      Object.entries(context).forEach(([key, value]) => {
        console.error(`   ${key}: ${value}`);
      });
    }

    // Log additional details
    if (details && details.length > 0) {
      const maxLines = ERROR_CONFIG.formatting.maxContextLines;
      const displayDetails = details.slice(0, maxLines);
      displayDetails.forEach(detail => console.error(`   ${detail}`));
      
      if (details.length > maxLines) {
        console.error(`   ... and ${details.length - maxLines} more details`);
      }
    }

    // Log stack trace in development
    if (ERROR_CONFIG.formatting.includeStackTrace && error && error.stack) {
      console.error('Stack trace:');
      console.error(error.stack);
    }
  }

  /**
   * Service-specific error logging with structured context
   * @param {string} serviceName - Name of the service (e.g., 'KeywordExtraction')
   * @param {string} operation - Operation that failed (e.g., 'extraction')
   * @param {Error} error - Error object
   * @param {Object} [context] - Additional context information
   */
  static logServiceError(serviceName, operation, error, context = {}) {
    this.logError({
      message: `${serviceName} ${operation} failed`,
      error,
      type: ERROR_TYPES.SERVICE_UNAVAILABLE,
      severity: ERROR_SEVERITY.HIGH,
      context: {
        service: serviceName,
        operation,
        ...context
      }
    });
  }

  /**
   * Application-level error logging (CLI, file operations, etc.)
   * @param {string} component - Application component (e.g., 'CLI', 'FileSystem')
   * @param {string} message - Error message
   * @param {Error} [error] - Original error object
   * @param {Array} [details] - Additional error details
   * @param {Object} [context] - Error context
   */
  static logAppError(component, message, error = null, details = [], context = {}) {
    this.logError({
      message: `${component}: ${message}`,
      error,
      type: ERROR_TYPES.INTERNAL_ERROR,
      severity: ERROR_SEVERITY.MEDIUM,
      details,
      context
    });
  }

  /**
   * Creates a standardized error result object
   * @param {boolean} success - Whether the operation was successful
   * @param {any} [data] - Success data (if success = true)
   * @param {string} [errorMessage] - Error message (if success = false)
   * @param {string} [errorType] - Error type from ERROR_TYPES
   * @param {Array} [details] - Additional error details
   * @returns {Object} Standardized result object
   */
  static createResult(success, data = null, errorMessage = null, errorType = ERROR_TYPES.UNKNOWN_ERROR, details = [], legacyErrorType = null) {
    if (success) {
      return {
        isValid: true,
        success: true,
        data
      };
    } else {
      const result = {
        isValid: false,
        success: false,
        error: errorMessage,
        errorType,
        details
      };
      
      // Add legacy error type for backward compatibility if provided
      if (legacyErrorType) {
        result.legacyErrorType = legacyErrorType;
      }
      
      return result;
    }
  }

  /**
   * Input validation helper with structured error reporting
   * @param {any} value - Value to validate
   * @param {Function} validator - Validation function that returns boolean or throws
   * @param {string} fieldName - Name of the field being validated
   * @param {Object} [options] - Validation options
   * @returns {Object} Validation result
   */
  static validateInput(value, validator, fieldName, options = {}) {
    try {
      const isValid = typeof validator === 'function' ? validator(value) : false;
      
      if (isValid) {
        return this.createResult(true, value);
      } else {
        const errorMessage = options.customMessage || `Invalid ${fieldName}`;
        return this.createResult(false, null, errorMessage, ERROR_TYPES.INVALID_INPUT, [
          `Field: ${fieldName}`,
          `Value: ${JSON.stringify(value)}`,
          ...(options.expectedFormat ? [`Expected: ${options.expectedFormat}`] : [])
        ]);
      }
    } catch (error) {
      return this.createResult(false, null, `Validation error for ${fieldName}`, ERROR_TYPES.INVALID_INPUT, [
        `Field: ${fieldName}`,
        `Error: ${error.message}`
      ]);
    }
  }

  /**
   * Assert that a required value is present and valid
   * @param {any} value - Value to check
   * @param {string} fieldName - Name of the required field
   * @throws {Error} If value is null, undefined, or empty string
   */
  static assertRequired(value, fieldName) {
    if (value === null || value === undefined || value === '') {
      const error = new Error(`Required field '${fieldName}' is missing or empty`);
      error.type = ERROR_TYPES.MISSING_REQUIRED_FIELD;
      error.field = fieldName;
      throw error;
    }
  }

  /**
   * Build file context information for error reporting
   * @param {string} filePath - Path to the file
   * @param {Object} [additionalInfo] - Additional file context
   * @returns {Object} File context object
   */
  static async buildFileContext(filePath, additionalInfo = {}) {
    const fs = await import('fs');
    const path = await import('path');
    
    const context = {
      type: CONTEXT_TYPES.FILE,
      filePath,
      fileName: path.basename(filePath),
      directory: path.dirname(filePath),
      ...additionalInfo
    };

    // Add file existence and stats if possible
    try {
      const exists = fs.existsSync(filePath);
      context.exists = exists;
      
      if (exists) {
        const stats = fs.statSync(filePath);
        context.size = stats.size;
        context.modified = stats.mtime.toISOString();
      }
    } catch (error) {
      // Ignore errors when building context
      context.contextError = 'Unable to read file stats';
    }

    return context;
  }

  /**
   * Build service context information for error reporting
   * @param {string} service - Service name
   * @param {string} operation - Operation name
   * @param {Object} [additionalInfo] - Additional service context
   * @returns {Object} Service context object
   */
  static buildServiceContext(service, operation, additionalInfo = {}) {
    return {
      type: CONTEXT_TYPES.SERVICE,
      service,
      operation,
      timestamp: new Date().toISOString(),
      ...additionalInfo
    };
  }

  /**
   * Build validation context for input validation errors
   * @param {string} fieldName - Name of the field being validated
   * @param {any} value - Value that failed validation
   * @param {string} [expectedFormat] - Expected format description
   * @returns {Object} Validation context object
   */
  static buildValidationContext(fieldName, value, expectedFormat = null) {
    return {
      type: CONTEXT_TYPES.VALIDATION,
      fieldName,
      providedValue: JSON.stringify(value),
      valueType: typeof value,
      ...(expectedFormat && { expectedFormat })
    };
  }

  /**
   * Update error handling configuration
   * @param {Object} newConfig - Configuration updates
   */
  static updateConfig(newConfig) {
    Object.assign(ERROR_CONFIG, newConfig);
  }

  /**
   * Get current error handling configuration
   * @returns {Object} Current configuration
   */
  static getConfig() {
    return { ...ERROR_CONFIG };
  }
}

export default ErrorHandler;
